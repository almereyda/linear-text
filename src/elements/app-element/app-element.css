.app-element {
  /* Dimensions include border and padding. */
  box-sizing: border-box;

  /* The app always fills the screen exactly and its contents overflow as
     needed. */
  width: 100vw;
  height: 100vh;
  overflow: auto;

  background: var(--color-light-grey);
}

.app-element__header {
  /* Fix the element to the viewport and remove from layout flow. */
  position: fixed;

  /* Raise above the content. */
  z-index: var(--z-index-header);
}
