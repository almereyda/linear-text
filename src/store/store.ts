import {Action, configureStore, ThunkAction} from '@reduxjs/toolkit'
import {tableSlice} from './table-slice/table-slice'
import undoable, {excludeAction, UndoableOptions} from 'redux-undo'

const undoConfig: Readonly<UndoableOptions> = Object.freeze({
  clearHistoryType: [tableSlice.actions.clearHistoryAction.type],
  filter: excludeAction([
    tableSlice.actions.addDraftAction.type,
    tableSlice.actions.focusAction.type
  ])
})

export function Store() {
  return configureStore({
    reducer: {table: undoable(tableSlice.reducer, undoConfig)}
  })
}

export const store = Store()

export type AppDispatch = typeof store.dispatch
export type RootState = ReturnType<typeof store.getState>
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>
