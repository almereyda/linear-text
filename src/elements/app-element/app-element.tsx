import {GroupElement} from '../group-element/group-element'
import {MenuCardElement} from '../menu-element/menu-element'
import {selectTableState} from '../../store/table-slice/table-slice'
import {useAppSelector} from '../../hooks/use-store'

import './app-element.css'

export type AppElementProps = Readonly<{'data-testid'?: string}>

export function AppElement({
  'data-testid': testID
}: AppElementProps): JSX.Element {
  const tableState = useAppSelector(selectTableState)
  return (
    <div className='app-element' data-testid={testID}>
      <header className='app-element__header'>
        <MenuCardElement />
      </header>
      <main className='app-element__body'>
        <GroupElement lines={tableState.table.lines} />
      </main>
    </div>
  )
}
