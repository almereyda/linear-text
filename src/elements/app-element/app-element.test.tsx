import {render, screen} from '@testing-library/react'
import {Provider} from 'react-redux'
import {store} from '../../store/store'
import {AppElement} from './app-element'

test('matches snapshot', () => {
  render(
    <Provider store={store}>
      <AppElement data-testid='app' />
    </Provider>
  )
  expect(screen.getByTestId('app')).toMatchSnapshot()
})
