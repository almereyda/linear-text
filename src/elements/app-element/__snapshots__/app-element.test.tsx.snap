// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="app"
  data-testid="app"
>
  <header
    class="app__header"
  >
    <div
      class="menu-card"
    >
      <div
        class="card "
      >
        <nav
          class="menu"
        >
          <ul
            class="list menu__list"
          >
            <li
              class="menu__list-item"
            >
              <div
                class="icon-button"
              >
                <button
                  accesskey="+"
                  class="buttonish "
                  title="Add new line [alt–+]."
                >
                  <img
                    alt=""
                    class="icon"
                    decoding="sync"
                    height="20"
                    loading="eager"
                    src="add-line-icon.svg"
                    width="20"
                  />
                </button>
                <label
                  class="icon-buttonish__label "
                >
                  Add
                </label>
              </div>
            </li>
            <li
              class="menu__list-item"
            >
              <div
                class="icon-button"
              >
                <button
                  accesskey="|"
                  class="buttonish "
                  title="Add group [alt–|]."
                >
                  <img
                    alt=""
                    class="icon"
                    decoding="sync"
                    height="20"
                    loading="eager"
                    src="add-group-icon.svg"
                    width="20"
                  />
                </button>
                <label
                  class="icon-buttonish__label "
                >
                  Group
                </label>
              </div>
            </li>
            <li
              class="menu__list-item"
            >
              <div
                class="icon-button"
              >
                <button
                  accesskey="-"
                  class="buttonish "
                  title="Remove selected line(s) [alt–-]."
                >
                  <img
                    alt=""
                    class="icon"
                    decoding="sync"
                    height="20"
                    loading="eager"
                    src="remove-line-icon.svg"
                    width="20"
                  />
                </button>
                <label
                  class="icon-buttonish__label "
                >
                  Discard
                </label>
              </div>
            </li>
            <li
              class="menu__list-item"
            >
              <div
                class="icon-button"
              >
                <button
                  accesskey="z"
                  class="buttonish "
                  title="Restore the previous state [alt–z]."
                >
                  <img
                    alt=""
                    class="icon"
                    decoding="sync"
                    height="20"
                    loading="eager"
                    src="undo-icon.svg"
                    width="20"
                  />
                </button>
                <label
                  class="icon-buttonish__label "
                >
                  Undo
                </label>
              </div>
            </li>
            <li
              class="menu__list-item"
            >
              <div
                class="icon-button"
              >
                <button
                  accesskey="y"
                  class="buttonish "
                  title="Reverse an undo [alt–y]."
                >
                  <img
                    alt=""
                    class="icon"
                    decoding="sync"
                    height="20"
                    loading="eager"
                    src="redo-icon.svg"
                    width="20"
                  />
                </button>
                <label
                  class="icon-buttonish__label "
                >
                  Redo
                </label>
              </div>
            </li>
            <li
              class="menu__list-item"
            >
              <div
                class="icon-button"
              >
                <button
                  accesskey="o"
                  class="buttonish "
                  title="Load an existing file [alt–o]."
                >
                  <img
                    alt=""
                    class="icon"
                    decoding="sync"
                    height="20"
                    loading="eager"
                    src="load-file-icon.svg"
                    width="20"
                  />
                </button>
                <label
                  class="icon-buttonish__label "
                >
                  Open…
                </label>
              </div>
            </li>
            <li
              class="menu__list-item"
            >
              <div
                class="icon-button"
              >
                <button
                  accesskey="s"
                  class="buttonish "
                  title="Download file [alt–s]."
                >
                  <img
                    alt=""
                    class="icon"
                    decoding="sync"
                    height="20"
                    loading="eager"
                    src="save-file-icon.svg"
                    width="20"
                  />
                </button>
                <label
                  class="icon-buttonish__label "
                >
                  Save
                </label>
              </div>
            </li>
            <li
              class="menu__list-item"
            >
              <div
                class="icon-button"
              >
                <button
                  class="buttonish "
                  title="Open save dialog."
                >
                  <img
                    alt=""
                    class="icon"
                    decoding="sync"
                    height="20"
                    loading="eager"
                    src="save-file-as-icon.svg"
                    width="20"
                  />
                </button>
                <label
                  class="icon-buttonish__label "
                >
                  Save As…
                </label>
              </div>
            </li>
            <li
              class="menu__list-item"
            >
              <div
                class="icon-button"
              >
                <button
                  accesskey="n"
                  class="buttonish "
                  title="Create an empty file [alt–n]."
                >
                  <img
                    alt=""
                    class="icon"
                    decoding="sync"
                    height="20"
                    loading="eager"
                    src="new-file-icon.svg"
                    width="20"
                  />
                </button>
                <label
                  class="icon-buttonish__label "
                >
                  New
                </label>
              </div>
            </li>
            <li
              class="menu__list-item"
            >
              <div
                class="icon-button"
              >
                <button
                  accesskey="?"
                  class="buttonish "
                  title="Open documentation [alt–?]."
                >
                  <img
                    alt=""
                    class="icon"
                    decoding="sync"
                    height="20"
                    loading="eager"
                    src="help-icon.svg"
                    width="20"
                  />
                </button>
                <label
                  class="icon-buttonish__label "
                >
                  Help
                </label>
              </div>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    class="app__main"
  >
    <article
      class="table"
    >
      <ul
        class="list table__list"
      />
    </article>
  </main>
</div>
`;
